bundles/org.openhab.binding.pegelonline/src/main/java/org/openhab/binding/pegelonline/internal/discovery/PegelDiscovery.java

```diff
@@ -0,0 +1,122 @@
+public class PegelDiscovery extends AbstractDiscoveryService implements DiscoveryService, ThingHandlerService {
+    private final Logger logger = LoggerFactory.getLogger(PegelDiscovery.class);
+    private Optional<PegelOnlineHandler> handler = Optional.empty();
+    private PointType homeLocation = UNDEF_LOCATION;
+    private HttpClientFactory httpClientFactory;
+
+    @Activate
+    public PegelDiscovery(final @Reference HttpClientFactory hcf, final @Reference LocationProvider lp) {
+        super(SUPPORTED_THING_TYPES_UIDS, 10, false);
+        httpClientFactory = hcf;
+        PointType location = lp.getLocation();
+        if (location != null) {
+            homeLocation = location;
+        } else {
+            logger.debug("No home location found");
+        }
+    }
+
+    @Override
+    protected void startScan() {
+        double homeLat = homeLocation.getLatitude().doubleValue();
+        double homeLon = homeLocation.getLongitude().doubleValue();
+        try {
+            ContentResponse cr = httpClientFactory.getCommonHttpClient().GET(STATIONS_URI);
+            Station[] stationArray = GSON.fromJson(cr.getContentAsString(), Station[].class);
+            if (stationArray != null) {
+                for (Station s : stationArray) {
+                    double distance = Utils.getDistanceFromLatLonInKm(homeLat, homeLon, s.latitude, s.longitude);
+                    if (distance < DISCOVERY_RADIUS) {
+                        logger.trace("Station in range {},{}", s.longname, s.water.shortname);
+                        reportResult(s);
+                    }
+                }
+            } else {
+                logger.trace("No stations found in discovery");
+            }
+        } catch (ExecutionException | TimeoutException | InterruptedException e) {
+            logger.trace("Exception during station discovery: {}", e.getMessage());
+        }
+    }
+
+    public void reportResult(Station s) {
+        String label = "Pegel Station " + Utils.toTitleCase(s.shortname) + " / " + Utils.toTitleCase(s.water.shortname);
+        Map<String, Object> properties = new HashMap<String, Object>();
+        properties.put("agency", s.agency);
+        properties.put("km", s.km);
+        properties.put("river", s.water.longname);
+        properties.put("station", s.longname);
+        properties.put("uuid", s.uuid);
+        properties.put("location", s.latitude + "," + s.longitude);
+        ThingUID uid = new ThingUID(STATION_THING, s.uuid);
+        thingDiscovered(DiscoveryResultBuilder.create(uid).withRepresentationProperty("uuid").withLabel(label)
+                .withProperties(properties).build());
+    }
+
+    @Override
+    public void deactivate() {
+        super.deactivate();
+    }
+
+    @Override
+    public void setThingHandler(ThingHandler h) {
+        if (h instanceof PegelOnlineHandler) {
+            handler = Optional.of((PegelOnlineHandler) h);
```

Comment:
```suggestion
    public void setThingHandler(ThingHandler thingHandler) {
        if (thingHandler instanceof PegelOnlineHandler pegelOnlineHandler) {
            handler = Optional.of(pegelOnlineHandler);
```

try to use meaningfull names and java 17 pattern matching preventing a double cast

Reply:
fixed